text/0000-localizing-api-messages.md

```diff
@@ -0,0 +1,135 @@
+only clients that are built using Stripes._**
+</span>.
+
+### Option 1
+Distribute them over the existing front-end modules:
+
+ui-checkin with translation key ui-checkin.mod-circulation.itemNotFound
+ui-requests with translation key ui-requests.mod-circulation.patronHasItemOnLoad
+
+*Pros*
+- Uses existing front-end modules
+
+*Cons*
+- Cannot disable a front-end module for a tenant if it hosts a mod-circulation translation key that is needed by another enabled front-end module.
+- Not all originating messages can be mapped to a frontend module 
+
+### Option 2a
+For each back-end module with translation keys create a dedicated front-end module with language files:
+
+For example create ui-mod-circulation module, and use translation keys
+
+ui-mod-circulation.itemNotFound
+ui-mod-circulation.patronHasItemOnLoad
+
+*Pros*
+- Any front-end module that uses mod-circulation can require ui-mod-circulation to load the translation files.
+- All mod-circulation translation strings go into a single module.
+*Cons*
+- Additional overhead in managing more ui modules
+- Not all originating messages can be mapped to a frontend module
+
+### Option 2b
+For each back-end interface with translation keys create a dedicated front-end module with language files:
+
+For example create i18n-circulation module, and use translation keys
+
+i18n-circulation.itemNotFound
+i18n-circulation.patronHasItemOnLoad
+
+*Pros*
+- Any front-end module that uses the circulation API can require i18n-circulation to load the translation files.
+- All circulation translation strings go into a single module.
+*Cons*
+- Additional overhead in managing more ui modules
+- Not all originating messages can be mapped to a frontend module
+
+### Option 3
+The back-end module hosts the language files.
+
+mod-circulation creates a translations/mod-circulation/ directory.
+
+Stripes fetches the translation files from the back-end module and can process them the same way as translation files from frone-end modules.
+
+*Pros*
+- Software and language files are in the same repository. Options 1 and 2 require to add the message string to some other repository using a second pull request.
+
+*Cons*
+- Is this technically feasible? How can Stripes fetch languages files from a non-Stripes module?
+
+### Option 4:
+When calling the back-end pass the required locale. The back-end maintains the translations files, replaces the placeholders and puts the final string into the "message" property. A lang query parameter that many FOLIO API interfaces already have or an "accept-language" header line might be used.
+
+*Pros*
+- Java code and translation files live in the same repository. No work for the front-end.
+- This is a well known pattern in the Java community
+
+*Cons*
+- None
+
+## Risks and Drawbacks
+
+None that we can think of. The pattern we are adopting is a standard pattern in the java community
+
+
+## Unresolved Questions
+- How do we allow tenants to change default locale ?
```

Comment:
Default locale is the locale that will be used when the locale information is not available when loading the resource bundles. For US tenants, I expect this to be English and for a German tenant it should be German. Without setting a default value for the locale, it will default to what the JDK is using.  Person installing the JRE should set the appropriate locale.

Reply:
> Default locale is the locale that will be used when the locale information is not available when loading the resource bundles.

My interpretation of this RFC is that the default locale (from the back end modules perspective) is always US English.

How is that default tenant specific?

> For US tenants, I expect this to be English and for a German tenant it should be German. Without setting a default value for the locale, it will default to what the JDK is using. Person installing the JRE should set the appropriate locale.

Oh, we are expecting that system operators change the configuration of the JRE to specify a locale.

This does not seem to be outlined in this RFC. Have I missed that?

As I understand it, that contradicts the principle that FOLIO module instances can be used by multiple tenants (each of which could operate a different locale).

Reply:
> My interpretation of this RFC is that the default locale (from the back end modules perspective) is always US English.

Agreed. A module should not have any knowledge of a tenant's locale or its JRE's locale, only a request's locale. My understanding is that a module's HTTP response should reflect the HTTP request's `Accept-language` header, or `en-US` if no such header is present.